### ISSUE TITLE
Having issues with gk-deploy script

### ISSUE THREAD

**Issue by mjschmidt** [gluster/gluster-kubernetes#434]
Hi I am running kubernetes 1.8.5 on centos7. 

I tried the script, but I get the glusterd not running. I seem to be having issues enabling the service, I checked some of the other issues on here, noticed that some other people were having similar problems so I went to abort my gk-deploy and the gluster pods did not delete. 

Question 1
Is that an issue or can I keep trying to now re run the gk-deploy script even though my gluster pods are still up?
If it is an issue how can I delete my gluster deamonset without causing and possible kubernetes caching problems?

Question 2
Why is it complaining about glusterd service not being started? Is that why I read that you needed to get rid of everything yum installed except for the gluster client?

Question 3
@jarrpa halp me. I see you answer a lot of these questions, getting exact error logs is tough because I am in an offline environment, but I will transcribe the important looking parts

**Comment by mjschmidt** [comment]
3rd one is admittedly not a question.

**Comment by jarrpa** [comment]
Hey there! Sorry to hear about your troubles. Let's see...

1. This is definitely an issue. While the abort function is somewhat incomplete, it should be getting rid of the GlusterFS Pods and DaemonSet. Just do a `kubectl delete ds` on the DaemonSet, that should also remove the pods. There shouldn't be any Kube caching problems. Be sure to delete any remaining GlusterFS configuration directories (e.g. /var/lib/glusterfs) on the host. You can get a list of these directories from the hostPath mounts in the GlusterFS DaemonSet template.
2. Could be any number of things. To start, make sure you have the necessary firewall ports open, the required kernel modules loaded, and that the glusterfs-server RPM is not installed.
3. Still a valid enough point. :) Can you take screenshots or photos of a terminal and post those? If not we'll have to make do.

Try a few of my recommendations above. If nothing works, then I'm going to want to look at:
* The output of `gk-deploy` with `-v` around when it fails
* Any relevant events in the output of `kubectl describe po` on one of the GlusterFS pods
* Any relevant-looking error (marked `E`) lines from the tail-end of `/var/log/glusterfs/glusterd.log` on any of the nodes running GlusterFS pods.

**Comment by mjschmidt** [comment]
How can I check that the correct kernels modules exist?

**Comment by mjschmidt** [comment]
I will not be able to take screenshots, but I can do my best to get you the most info that I can.

**Comment by mjschmidt** [comment]
I found the host path mounts, so I will clear those out. For some weird reason the script deployed two identical deamonsets deployments to kubernetes. It still only deployed a single set of pods, but that may have been why the script failed to take down the Deamonset on --abort.

**Comment by jarrpa** [comment]
If you ran the script more than once that would account for why there are more than one DaemonSets... that's the only thing I can think of, anyway. But the presence of another set of GlusterFS pods would account for some of these issues. Make sure the DaemonSets are removed and any associated node labels are also removed.

For kernel modules, `lsmod | grep <name>` will show you if a given module is present, and `modprobe <name>` will load a given module. You'll want to create a file in `/etc/modprobe.d/` to load those modules at startup would also be a good idea.

**Comment by mjschmidt** [comment]
Sorry I am looking at the setup guide and I don't see the modules that I need there?

For the host paths, should I delete the folder as well. For example should I remove the whole heketi folder?

I noticed in /run/lvm/ there are two files in there, I went to look at them and I got permission denied even as root. Is that lvm folder created by the gluster deployment (and therefor I can delete) or could it be created by something else and I shouldn't just recursively delete that folder?

In the folder I see a lvmetad.socket and a lvmpolld.socket file

**Comment by jarrpa** [comment]
NOnono, do not delete any system directories! Only delete the ones specific to GlusterFS and heketi (they should be obvious by name). And yes, you can remove the directories themselves.

If you don't skip it, the intro text to gk-deploy tells you what you need to have available: https://github.com/gluster/gluster-kubernetes/blob/master/deploy/gk-deploy#L528-L560

**Comment by mjschmidt** [comment]
I already did... just kidding ( : 
So I plan to delete these folders:
/var/lib/heketi
/etc/glusterfs
/var/log/glusterfs (though I cant see why logs would matter)
/var/lib/glusterd
/var/lib/misc/glusterfsd

and not delete:
/run/lvm
/dev (obviously)
/sys/fs/cgroup
/etc/ssl

Ah I see in the gk-ddeploy skip the kernel modules. I can look at those next. Let me know if proposed delete/donotdelete list looks reasonable. Thanks!

**Comment by jarrpa** [comment]
A reasonable list. Go ahead!

**Comment by mjschmidt** [comment]
Okay I got my modules loaded up, I had seen that before but for some reason choose not to explore it more. Those are all loaded up now.

how can I check that mount.glusterfs is available?

**Comment by mjschmidt** [comment]
Can I just sudo yum install the glusterfs-fuse.x86_64 package on centos7 and assume it is there?

**Comment by mjschmidt** [comment]
list of packages proposed to yum install on worker nodes:
centos-release-gluster313 (or 312)
glusterfs-cli.x86_64
glusterfs-fuse.x86_64

**Comment by mjschmidt** [comment]
I decided for now I would only get the cli and fuse packages I listed above and give it a try. 

The deployment was not successful
Error:
unable to create new node: New Node doesn't have glusterd running
How do I turn on glusterd?

I saw that you can enable glusterd.service , but when I tried that I got no such file or directory.

**Comment by jarrpa** [comment]
If you're running GlusterFS in pods, heketi is looking for clusterd in the contaiers and there must be NO glusterd present on the host. You do not "turn on" glusterd in this case, it should be running when the containers come up. The error you're getting from heketi could mean a number of things, but fundamentally it means it can't talk to a glusterd process where it expects it should. Check the outputs of `kubectl logs <heketi_pod>` and `kubectl describe po <gluster_pod>`. If nothing shows there, check the GlusterFS logs on the hosts as I mentioned above.

`glusterfs-fuse` is the only RPM you need to install to mount GlusterFS volumes.

**Comment by mjschmidt** [comment]
Okay got it, I will remove the gluster-cli from the nodes then

**Comment by mjschmidt** [comment]
By the way, the gk-deploy is still creating 2 deamonset deployments in kubernetes, but only one set of pods (so just when I go to tear it down I have to delete the deamonset manually and clear out the host file folders, which isn't a huge deal)

So things running:
the three modules
gluster-fuse.x86_64

Still getting the unable to create node from heketi, so I'll grab the logs and transcribe them.

kubectl logs heketipod:
Error <time is here> /src/github.com/heketi/heketi/executors/kubeexec/kubeexec.go:301: Get https://10.96.0.1:443/api/v1/namespace/default/pods?labelSelector=glusterfs-node: dial tcp 10.96.0.1:443 i/o timeout
And then goes on to say that it fails to get list of pods

kubectl describe po gluster pod:
I don't know what you would be interested in here?

I see a Tolerations section:
node.aplha.kubernetes.io/notReady:NoExecute
node.aplha.kubernetes.io/unreachable:NoExecute
node.kubernetes.io/disk-pressure:NoSchedule
node.kubernetes.io/memory-pressure:NoSchedule

**Comment by mjschmidt** [comment]
When I describe the gluster pod it says that it is initialized, ready and scheduled as True

**Comment by mjschmidt** [comment]
From the heketi logs I also see:
Started GET /clusters/<long id number>
Completed 200 OK in <time>
Started delete /clusters/<same long id number>
<Two info statements about cluster being deleted here>
ERROR <time> /src/github.com/heketi/heketi/apps/glusterfs/allocator_simple.go: Unknown cluster id requested

**Comment by jarrpa** [comment]
Please don't spam multiple messages if you can avoid it.

On the describe I'm looking for Events towards the bottom to see if any of them indicate an error.

Are you sure you have no GlusterFS DaemonSets or Pods running after you clean up a failed deployment? Are you using something other than the default namespace and could there be something conflicting in another namespace?

You never answered an earlier question: do you have the necessary firewall ports open on all GlusterFS nodes?

**Comment by mjschmidt** [comment]
> Please don't spam multiple messages if you can avoid it.
Sure not a problem

For firewall ports I have a security group set up that allows:
TCP over ports 2222, 24007, 24008, 49149-49251 

So that shouldn't be an issue. Also I think with our setup if nodes are in the same region then communication within the region should not be restricted... I just set up the security group to be extra safe that ports would not be an issue.

I am sure that I clean up all the pods, first I delete the deamonset with: 
kubectl delete ds/glusterfs 

by the way on that deamonset note, its still doing the weird thing where it creates two deamonsets but deploys one set of pods. I have seen that before on other issues, but I don't know if it was ever resolved.

Then I do a kubectl get all to make sure nothing is there and I can see only the svc/kubernetes service running.

I went to the host machine and noticed that when I did a sudo yum list installed | grep gluster 
that a few packages were installed in addition to fuse as dependecies:
glusterfs.x86_64                                                             3.8.4-18.4.e17.centos    @os_centos_base
glusterfs-client-xlators.x86_64                                       3.8.4-18.4.e17.centos    @os_centos_base
the one I wanted-> glusterfs-fuse.x86_64                     3.8.4-18.4.e17.centos    @os_centos_base
glusterfs-libs.x86_64                                                      3.8.4-18.4.e17.centos    @os_centos_base

Is this an issue?

**Comment by jarrpa** [comment]
Hmm... yes, the networkign you described should be working.

I don't recall other issues with multiple DaemonSets... what are the names of the DaemonSets and are they in the same Namespace?

"all" may not get everything we're interested in. Try `kubectl get deployment,ds,po,job,secret,ep,svc,sa --all-namespaces` and see if there's anything with the words "glusterfs" or "heketi" in it.

Those packages are the correct ones. The reason I only specify `glusterfs-fuse` is that it takes care of installing all needed dependencies.

**Comment by mjschmidt** [comment]
Hmmmm insteresting, so after the abort I did a kubectl get all (in just my namespace) and I see the deamonset twice, however when I do a kubectl get ds (Just in my namespace at first) I only see the one deamonset. So this may be an issue with the kubectl get all. As a result, I think that can be ignored for now.

kubectl get deployment,ds,po,job,secret,ep,svc,sa --all-namespaces | grep gluster 

and only saw the pods and the daemonset deployment running, then I deleted the deamonset with 
kubectl delete ds glusterfs and that deleted out just fine

kubectl get deployment,ds,po,job,secret,ep,svc,sa --all-namespaces | grep gluster and found nothing
kubectl get deployment,ds,po,job,secret,ep,svc,sa --all-namespaces | grep heketi also found nothing

Other things I can check? Oh by the way the names of the deamonsets when I had two of them (but again only one set of pods) was ds/glusterfs and ds/glusterfs, so they were named the same.

**Comment by jarrpa** [comment]
OKay yeah something is screwy with your cluster. You can't have two resources of the same name and type in the same namespace.

The command you listed is definitely fine.

If you're at a point where the heketi and GlusterFS pods are up and running but heketi can't communicate to GlusterFS, check that all pods are showing `1/1 Ready`. Next try going into a GlusterFS container and running `systemctl status glusterd` to see if the process is running. If all this is fine, you may have something weird in your networking. Note that the GlusterFS pods use host networking, meaning that they assume the network identity of the hosts they're running on. You would have to verify that processes in containers in your cluster can find routes to the host IP addresses.

**Comment by mjschmidt** [comment]
glusterd is running in the container

so that is good. 

I ran hostname -i from inside the container and that worked and I got the correct IP back from the command.

Any test commands I can run from inside the gluster container that will let me know "yes the cluster is connected"? I did a gluster peer probe to one of the other containers and that worked. Then I did a gluster peer status and saw there was only one peer (the one I had just connected to) so it seems like that isn't quite working correctly? 

I went into the heketi container as well and did a hostname -i and that returned command not found. What can I do from that container to see if things are working correctly or not as I debug more?

**Comment by mjschmidt** [comment]
I have a 5 node cluster though. 

What I am saying is that the gluster pods are not connecting to each other via the script because when I go into a gluster container origonally the container has 0 peers, however when I manually add a peer in the container that works and it connects to a different host ip.

busybox commands and values:
hostname = busybox-<long thing>
hostname -i = 10.244.5.3 (this is an ip of the container)
ping 10.2.46.33 (one of the kubernetes workers) = no packets transmitted or returned from busybox container
ping 10.111.170.255 (kubernetes service ip for heketi) = nothing
ping 10.111.170.255:8080 (same ip but now over port 8080) = nothing

**Comment by mjschmidt** [comment]
So it looks like my busybox can't ping anything but my gluster container can ping anyt machine in the cluster

**Comment by mjschmidt** [comment]
I was looking at the gluster yaml file and noticed that there are not any ports specified on there? Is that expected? How does communication happen if there are no ports?

**Comment by mjschmidt** [comment]
Update, mostly so people who come after me can follow some sort of documentation

I was able to curl heketi and get heketi to say hello to me from inside the heketi pod going through the kubernetes heketi deployment service ip. the command is curl <service_ip for heketi-deploy>:8080/hello

I was also able to curl hello heketi from a single gluster container that is on the same node. So there are some networking issues going on.

deploy-heketi's ip is 10.111.174.68 with port 8080 open

I think if port 8080 was not open I wouldn't be able to do this... the question is, why does this service ip only work on the single node...

**Comment by jarrpa** [comment]
A few things:
* When you're in a container that is not using host networking, you want to verify two things:
  1. Can I access other pods via their service IPs? This is typically done via `curl` or `telnet`. This will tell you if cluster networking is functional.
  2. Can I access Kubernetes nodes via their host IPs? `ping` will suffice here. This will tell you if the cluster network has access to the network of the hosts.
  * Since GlusterFS runs with host networking, both of these must be true in order for it to function. The GlusterFS pods will all automatically be able to communicate with each other because they are on the same host network. The heketi pod is in the cluster network, and thus will need a route to access the host network.
* The script is failing because it is trying to communicate to the GlusterFS pods from the heketi pod.
* Since heketi is not using host networking, port 8080 does not have to be open on the host. Indeed, there is nothing needed in Kubernetes for a service to be able to receive traffic on any arbitrary port.
* Specifying network ports for a pod is not required, and in fact offers no real utility other than letting someone reading the YAML know what ports may be in use. Kubernetes does no checking or enforcement of network port usage.
* Given the way Kubernetes handles routing tables on the nodes, it makes sense that a pod on a given node would be able to ping the IP of that node.
* Documentation for interacting with heketi can be found at the project's Github repo. For this specific case, see here: https://github.com/heketi/heketi/blob/master/doc/admin/server.md#testing-the-configuration

**Comment by jmogden** [comment]
@jarrpa I'm having a similar issue to mjschmidt. You mentioned making a file in /etc/modprobe.d/ if I need to use `modprobe <name>` to load a module at start up; how do set that file up? Is there a example somewhere I can reference?

**Comment by jmogden** [comment]
@jarrpa Next question, which yaml file sets up a pod that gets labelled "heketi-storage-copy-job"?

**Comment by mjschmidt** [comment]
An update I am still working on this. I have learned more about iptables and networking than I anticipated. For anyone interested in a brief overview of IP tables role in Kubernetes the general gist is:

Kubernetes does a ton with IPtables...

**Comment by mjschmidt** [comment]
By the way it looks like it was issues with the way I was deploying my kubernetes cluster and flannel or docker networking issues. So the next question is now that my heketi has hooked up with my gluster containers I see it is trying to do the setup-openshift-heketi-storage.

I am not running on Openshift and don't need this. I am running on centos7 machines. How can I exclude openshift stuff from my deployment?

**Comment by nixpanic** [comment]
On Fedora/CentOS/RHEL systems you can drop an executabl file under
/etc/sysconfig/modules/ . An example would be:

    #!/bin/sh
    #
    # Force loading some modules
    
    grep -q -w dm_thin_pool /proc/modules || modprobe dm_thin_pool

Save that file as /etc/sysconfig/modules/gluster-bricks.modules and make
it executable. Upon a reboot the script should get executed and the
modules get loaded.

The configuration files in /etc/modprobe.d/ can be used to modify the
parameters while the module gets loaded. I dont think there is a way to
mark modules to get loaded automatically there.

**Comment by mjschmidt** [comment]
@nixpanic  So I am not very familiar with the modules, or etc/sysconfig. 

Does this look like it would be right?

These are the modules I need:
 dm_snapshot
 dm_mirror
 dm_thin_pool

Saving a file here /etc/sysconfig/modules/gluster-bricks.modules that contains:

    #!/bin/sh
    #
    # Force loading some modules

    grep -q -w dm_thin_pool /proc/modules || modprobe dm_thin_pool
    grep -q -w dm_thin_pool /proc/modules || modprobe dm_mirror
    grep -q -w dm_thin_pool /proc/modules || modprobe dm_snapshot
    
^End file
Then I will do a chmod +x /etc/sysconfig/modules/gluster-bricks.modules

Then reboot all the machines that I do this to (in this case the 3 kubernetes nodes that I plan to get this gluster running on).

Then run gk-deploy and profit?

**Comment by nixpanic** [comment]
@mjschmidt yes, that looks correct. But you need to update the `grep` commands in the script ;-)

But, maybe there is a better way. Sometimes kernels compile certain modules in their main executable, and do not offer to load them. In that case, `/proc/modules` will not list them either. To make it less dependent on the kernel/distribution, I should have suggested:

```
#!/bin/sh
#
# Force loading some modules

[ -d /sys/module/dm_thin_pool ] || modprobe dm_thin_pool
[ -d /sys/module/dm_mirror ] || modprobe dm_mirror
[ -d /sys/module/dm_snapshot ] || modprobe dm_snapshot
```

Try the script+reboot on one of your nodes, and verify that the modules are loaded (check in `/proc/module` or use `lsmod`).

**Comment by mjschmidt** [comment]
I will use the lsmod and grep for the desired modules ( : I'll get back to you in a bit.

**Comment by mjschmidt** [comment]
Okay cool it looks like that works ( : and that wasn't too bad, but what does that have to do with the heketi openshift stuff I was referring to earlier?

I don't think I am understanding that yet.

**Comment by nixpanic** [comment]
The `setup-openshift-heketi-storage` command executed by Heketi will create a Gluster Volume to store the heketi database on. Creating the LVM/LVs can fail if the `dm_*` modules are not loaded. These should get loaded automatically, but that only works if the container can load them (which might not be the case because of several reasons).

**Comment by jarrpa** [comment]
@mjschmidt @nixpanic I'll note that the `setup-openshift-heketi-storage` command is mislabeled. It applied to both Kubernetes and OpenShift. It is currently needed by the configuration used in `gk-deploy`, as it allows for robust persistent storage for the heketi database.

Supposedly there is a way that you can just specify module names in a flat file so that they get loaded on system startup. See here: https://github.com/mjudeikis/openshift-ansible/blob/7d67afffe217e29c7bb8941343bd017907ffc9df/roles/openshift_storage_glusterfs/tasks/kernel_modules.yml and https://github.com/mjudeikis/openshift-ansible/blob/7d67afffe217e29c7bb8941343bd017907ffc9df/roles/openshift_storage_glusterfs/templates/glusterfs.conf.j2

The format is just the names of modules, one per line.

**Comment by jarrpa** [comment]
@mjschmidt Yes, the `setup-openshift-heketi-storage` command by default will always create a YAML manifest for the copy job that specifies `heketi/heketi:dev` as its image. You can change that by using the `--image` option into the command. So you have two options:
1. Edit your gk-deploy script to add `--image <image_name>` here: https://github.com/gluster/gluster-kubernetes/blob/master/deploy/gk-deploy#L859
2. Pull the `heketi/heketi:dev` image or tag your existing image with that name.

**Comment by mjschmidt** [comment]
Hmmmm okay, I am going to try that and if it works tomorrow. Thank you.

**Comment by mjschmidt** [comment]
@jarrpa @nixpanic WHOOP

So good news after the hard code into the open-shift command (a change I very much want to open a new issue for/help out with when this is all said and done)  and countless cluster deletes and re-provisions, I was able to have the gk-deploy script run successfully. 


**NOW** how do I use this storage? 

Upon completion the script says "heketi is now running and accessible via http://<IP_here>:8080 to run admin command you can install heketi-cli and use it as follows:

etc
etc
etc

I am trying to have the storage I just set up serve the kubernetes cluster it is currently running on. Is that where that yaml file comes into play? So Throw the following into a yaml file then kubectl apply -f whatever_I_call_this file.yaml:

---
apiversion: storage.k8s.io/v1beta
kind: StorageClass
metadata:
  name: glusterfs-storage
provisioner: kubernetes.io/glusterfs
parameters:
  resturl: "http:<IP_HERE>:8080"


By the way is this IP generated or statically written and I have to fill in my IP? It looks like the later but I want to be sure.

**Comment by mjschmidt** [comment]
Never mind I just had to rename the volume mount point to the storage class I made. WOOOOOOO IT WORKS! 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

I'll have to find some time to create a troubleshooting page this weekend for Kubernetes deployment

So can I do the heketi deployment and scale that out if I need to and have requests come into the hekeiti in say a round robin fashion? Or would that start to break things?

**Comment by mjschmidt** [comment]
Not to spam the chat, but how do I expand out to more nodes?

**Comment by jarrpa** [comment]
WOO!! Great to hear! :D 

All right, so, at this point you don't want to scale the current heketi deployment. heketi expects exclusive access to its metadata DB. But there are several ways you can expand this setup:

1. Add more nodes to the current clusters. This can be done via the heketi-cli by executing `heketi-cli node add` with the node info, then add the devices on that node.
2. Add more clusters to the current heketi instance. heketi can manage multiple clusters from the same instance. I believe this is done by specifying a new `clusterID` in the `node add` command. By default heketi will just use that additional cluster in its storage allocation algorithm, but you can now also specify one or more clusterIDs in your StorageClasses to direct volume requests to specific clusters.
3. You can spin up a new heketi instance, either in a different namespace (easiest) or with a different name. You would also need a new GlusterFS DaemonSet. It would basically be running the gk-deploy process again just with different names.

**Comment by mjschmidt** [comment]
So adding gluster nodes to the deamon set is a manual process? I know if I label a node with the gluster tag it will spin up another gluster pod, but I guess there isn't anything that will trigger heketi to work with that new gluster pod?

**Comment by jarrpa** [comment]
Correct. We are discussing ways to do that, but it's still some time out.

**Comment by mjschmidt** [comment]
Okay can you give an example of the heketi node add command? 

Is there a way to hook my terminal to the heketi container if I install the heketi-cli on it, or do I have to bash into the container in order to do these commands?

**Comment by jarrpa** [comment]
You need to open a shell in the container to use the built-in `heketi-cli`, yes. You can also install `heketi-cli` on any machine that can send http requests to the heketi service. So if `curl://<heketi_url>/hello` works you can use `heketi-cli` from there. Regardless, you can run `heketi-cli -s http://localhost:8080 --user admin [--key <SECRET>] node add --help` to get the usage message. Note that the `SECRET` is only required if you configured one.

**Comment by mjschmidt** [comment]
Okay so I have the issue where a node contains devices, but the node is dead so I can't delete if from the heketi-cli but I also cant clean up any of my pvs that are deleted because one of the peers are missing.

How do I deal with this?

**Comment by mjschmidt** [comment]
When I try to disable and remove the node that fails because now the node is dead/gone

**Comment by mjschmidt** [comment]
@nixpanic any idea? Any idea of who I can ask?

**Comment by mjschmidt** [comment]
last try (and tomorrow morn) to reach @jarrpa or @nixpanic or anyone else before I try opening a new ticket (this could potentially be a new ticket anyway). 

I have glusterfs pretty much up and running great, and I have learned a lot about trouble shooting kubernetes nodes to bring them back to life. However, I want to know how deal with node death in this setup because heketi doesn't want to remove nodes that it thinks has bricks on it, but kubernetes doesn't want to clean up pvcs unless the whole gluster cluster is present, in turn causing the k8s cluster to run out of storage. What do I do if I am running gluster in k8s and a node literally blows up and is disconnected from the cluster with no hope of ever coming back. Lets also say in this hypothetical instance that I have 12 gluster nodes so I would still be above the 3 min requirement.

**Comment by jarrpa** [comment]
Looks like this conversation continued in https://github.com/heketi/heketi/issues/1120.

**Comment by mjschmidt** [comment]
It is I am still waiting on an answer from them on some stuff, is there a command I can run on gluster itself to see if it is out of space or not? My total pvcs do not take up all the space I provided, so I think it may be an issue with stuff not being cleaned up correctly?)

**Comment by jarrpa** [comment]
From gluster, no. From the GlusterFS pods you can run standard LVM command to inspect the configuration and usage. Ones to be likely be most relevant are `pvs`, `vgs`, `pvdisplay`, and `vgdisplay`.

**Comment by mjschmidt** [comment]
I did the lvm vgdisplay on the gluster node and I see what I think is the relevant volume (~100Gig)

I see 
VG size 99.87 gib
PE size 4.0 mib
total PE 225567
alloc pe / siz 23437 / 91.55 gib
Free PE 2130 / 8.32 gib
VG UUID AcF6yadayadayada

**Comment by mjschmidt** [comment]
Okay something between heketi and gluster isn't working right. Heketi I show my 14 volumes but when I look at the available space on the node I see no free space. This happens when ever I have gluster and heketi up for an extended period of time making believe that heketi isn't actually cleaning up the space it says it is. How do I ask gluster these questions, for example. what are all the volume id's you have? Because then I could compare volume id's in gluster with volume ids in heketi and see which volumes in gluster need destroying.

**Comment by jarrpa** [comment]
You may be hitting another heketi issue. You want to compare against the output of `gluster volume list`. `gluster volume status` will also tell you the names of each volume's bricks.

**Comment by mjschmidt** [comment]
great thanks @jarrpa

**Comment by mjschmidt** [comment]
I NOW KNOW WHAT THE PROBLEM IS! Wow that was a long road. Long story short this is a bug between gluster and heketi running in containers. My second pvc claim was for 1 brick with 3 replication according to heketi-cli, but somewhere in translation gluster is creating 8 bricks x 3 replication for a total of 24 bricks. At 11 gib each that is what is eating all my space.

Here is the long story if interested:
Okay so I found something very interesting that probably explains my space problems...
So I started out by checking out the gluster list and then checking out the volume info on each other the volumes. I found one that was creating 24 bricks! This pvc happened to be 11 gib each in size. That would explain my space problem. It also sees a second set pvc associated with the same stateful set deployment, but this set of pvcs do not have any problems associated with them so we ignore them for now and just know that I should be taking up and regaining a total of (11 gib for first claim per container * 3 for replication * 3 containers each making a pvc claim + 6 gib for first claim per container * 3 for replication * 3 containers each making a pvc claim = 153 total gib for the stateful set) Since in this stateful set deployment only had two containers deploy/attach pvcs I only had 101 gib that should be returning to a heketi that believes it is full.

So there is a disconnect between heketi and gluster somewhere. When I look at the volume in heketi using heketi-cli volume info <name> here are the important rows:
Size: 11
Mount: <ip>:vol_<vol_blah851b>
Durability type: replicate
Distributed+Replica: 3

When I bash into a gluster container and do gluster volume info <vol_blah851b> I see:
Volume ID: Distributed-Replicate
Volume ID: 6ba8blah-blah-blah-blah12aab7a3
Status: Started
Snapshot Count: 0
Number of Bricks: 8 x 3 = 24

This is where I believe the problem lies, a replica pod that has an identical pvc claim is only getting 1 x 3

I cleared out those pvcs from kubernetes to see what would happen and to make sure my stuff would get cleaned up.
Here is what I found:
I take down the statefulset that had the associated pvc claims
After the stateful set is gone I clean up the PVCs with kubectl delete pvc <pvc 1> <pvc 2> <pvc 3>
I check the heketi-cli to make sure the 11 gib volumes I expect to be deleted are deleted:
I check gluster to make sure that the associated volumes are deleted and they were in fact deleted

I then went back to heketi-cli and did some volume node info to make sure I gained back my space, but since there is a disconnect between what gluster thinks is happening and what heketi thinks is happening I only have 108 gib of space total even though I should have gotten back 264 gigs of space for the one volume that . Heketi believes that when it cleared out the pvcs it regained

**Comment by mjschmidt** [comment]
https://github.com/heketi/heketi/issues/1127

also posted here

**Comment by jarrpa** [comment]
....dang. That's really weird, I've never heard of heketi creating an 8x3 volume from a dynamic provisioning request. Are you able to reproduce this behavior consistently?

We have also been aware of heketi's tendency to get out of sync with the state of Gluster underneath it. The most recent release of heketi does a lot to address these issue. Check if you're running heketi v6.

**Comment by mjschmidt** [comment]
I was not before, I just switched to heketi:6 today on a seperate cluster for testing, but now I am having deploy issues I am working through for that. I will update on that front when I figure out the nitty gritty details. My question on that front is, can I run heketi 6 with older gluster containers (what ever the latest was when we first started this thread and weren't best friends)
Yes the heketi database seems to be getting out of sync with gluster underneath I saw the stuff about heketi 6. I really hope that fixes things 
In general, am I able to resync the database with gluster? How do I do that?

**Comment by jarrpa** [comment]
You should be able to run heketi:6 with older GlusterFS images, yes. And yes, it is possible with some new DB management tools to tyr and repair an out-of-sync DB. Ask in the heketi project for more info on that.

**Comment by mjschmidt** [comment]
I am putting this at the top, but its really the last thing you should read:
OH MY GOD I DID ALL THIS TRANSCRIBING TO FIND OUT GLUSTERFS DELETED A VOLUME AND HEKETI DIDN'T KEEP TRACK (I did heketi-cli volume list and got that the volume was not cleaned up then did a gluster volume list and saw gluster did in fact clean it up... fuq...)

Okay I deployed heketi:6 to our old cluster, the new cluster I was setting up was being done with puppet and there was no way to see if it was a puppet issue or a heketi issue. I have confirmed it is a puppet issue, so lets ignore that for now and we are staying consistent by staying with the old cluster. I deployed heketi:6 successfully with "old" (its not that old) gluster and successfully created a dynamic pvc claim.

However now when I delete the pvc claim the pv is not cleaning up. I bashed into the hekti container and 
Please keep in mind I have to hand transcribe these logs, so I am getting the relevant stuff. and ignoring stuff like ids and times
I see the following relevant information from the logs:
[negroni] Started DELETE /volumes/4cdaa...
[heketi] INFO 2018/04/25 Loaded simple allocator
[negroni] Completed 202 Accepted in 23.03ms
[asynchttp] Info <time and data> Start job d6444r8...
[heketi] INFO 2018/04/25 Started asymc operation: Delete Volume
[negroni] Started GET /queue/d6444r8...
[negroni] Completed 200 OK in 85.9
[kubeexec] DEBUG 2018/04/25 /src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker7 Pod: glusterfs-vnzql --mode=script snapshot list vol_4cdaa... --xml
Result: <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<cliOutput>
  <opRet>0</opRet>
  <opErrno>0</opErrno>
  <opErrstr/>
   <snapList>
        <count>0</count>
    </snapList>
</cliOutput>
[kubeexec] DEBUG 2018/04/25 /src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker7 Pod: glusterfs-vnzql --Command lvs --options=lv_name, thin_count --separator=:
Result:      LV:#Thins
  vol_data:
  brick_552c...:
  tp_552c...:1
  lv_home:
  lv_home:
  lv_root:
  lv_swap:
  lv_temp:
  lv_usr:
  lv_usr:
  lv_var:
  lv_var:
[kubeexec] DEBUG 2018/04/25 /src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker5 Pod: glusterfs-2g16b Command: lvs --options=lv_name, thin_count --separator=:
Result:      LV:#Thins
  vol_data:
  brick_b9c6...:
  tp_b9c6..:1
  lv_home:
  lv_home:
  lv_root:
  lv_swap:
  lv_temp:
  lv_usr:
  lv_usr:
  lv_var:
  lv_var:
[kubeexec] DEBUG 2018/04/25 /src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker4 Pod: glusterfs-pg6ct Command: lvs --options=lv_name, thin_count --separator=:
Result:      LV:#Thins
  vol_data:
  brick_d0d4...:
  tp_d0d4..:1
  lv_home:
  lv_home:
  lv_root:
  lv_swap:
  lv_temp:
  lv_usr:
  lv_usr:
  lv_var:
  lv_var:
[negroni] Started GET /queue/d644...
[negroni] Completed 200 OK in 181ms
[negroni] Started GET /queue/d644...
[negroni] Completed 200 OK in 73ms
[negroni] Started GET /queue/d644...
[negroni] Completed 200 OK in 73ms
[negroni] Started GET /queue/d644...
[negroni] Completed 200 OK in 73ms
[kubeexec] DEBUG 2018/04/25 /src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host:
worker7 Pod: glusterfs-vnzql  Command: gluster --mode=script volume stop vol_4cdaa... force
[negroni] Started GET /queue/d644...
[negroni] Completed 200 OK in 73ms
Result volume delete: vol_4cdaa...: success
[negroni] Started GET /queue/d644...
[negroni] Completed 200 OK
[kubeexec] DEBUG src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker7 Pod: glusterfs-vnzql Command: umount /var/lib/heketi/mounts/vg_7a90.../brick_d0d4...
Result: 
[kubeexec] DEBUG src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker5 Pod: glusterfs-vnzql Command: umount /var/lib/heketi/mounts/vg_548f.../brick_552c...
Result: 
[kubeexec] DEBUG src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker4 Pod: glusterfs-vnzql Command: umount /var/lib/heketi/mounts/vg_357e.../brick_b9c6...
Result:
[negroni] Started GET /queue/d644
[negroni] Completed 200 OK
[negroni] Started GET /queue/d644
[negroni] Completed 200 OK
[negroni] Started GET /queue/d644
[negroni] Completed 200 OK
[negroni] Started GET /queue/d644
[negroni] Completed 200 OK
[negroni] Started GET /queue/d644
[negroni] Completed 200 OK
[negroni] Started GET /queue/d644
[negroni] Completed 200 OK

**Comment by mjschmidt** [comment]
Also @jarrpa I aborted the gluster deployment and I am left over with a brick that was not cleaned up.

I am as a result unable to delete my volume because it is still in use by another device.

A lvc --segments reveals
LV                                               VG                                       Attr                   #str Type           ssize
vol_data                                    vg01                                    -wi-ao----             1 liniar              297g
brick_39<id>                             vg_ee..                                 Vwi-a-tz--             0thin              2.00g
tp_39...                                       vgee                                     twi-aotz--            1 thin_pool      2.00g

I want to destroy this brick without gluster (since gluster deploy is down) so that I can ultimately destroy my lvm

**Comment by jarrpa** [comment]
`wipefs -a` on the underlying device should do the trick.

**Comment by mjschmidt** [comment]
Okay I already deleted my volumes but if I encounter this again its good to know how to handle it (:

**Comment by mjschmidt** [comment]
@jarrpa have you seen this issue with heketi:6 container? From the logs it looks like heketi is starting to do all the right things then it gets stuck then never retries to continue onward? Thoughts? I don't see any gluster issues from the heketi logs.

**Comment by jarrpa** [comment]
I'm not sure, I don't track exact heketi versions as well as I should. I know in the 6 there was a lot of work put into improving stability especially with regards to volume creation and deletion. Is the current issue that you can't get the volumes to delete? Is this via kubectl, heketi-cli, or both?

**Comment by mjschmidt** [comment]
Do you guys run containerized gluster and heketi? Or is this something that is not really being done?

The problem is that it seems like heketi is getting stuck provisioning volumes in the middle of creation then freezing up.

I see this in the heketi logs
```
[kubeexec] DEBUG src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker7 Pod: glusterfs-vnzql Command: lvcreate --poolmetadatasize 16384K -c 256K -L 3145728K -T vg_548f.../tp8f08... -V 31457287K -n brick_8f08...
Result: Using defult stripesize 64.00 KiB
Thin pull volume with chunk size 256 Kib can address at most 63.25 Tib of data
Logical volume "brick_8f08..." created
I am tired of typing, but the point is that it gets to this point
[kubeexec] src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker7 Pod: gluster-vnzql Command: mount -o rw,inode64,nouuid /dev/mapper/vg_548f...
Result:
[kubeexec] src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker7 Pod: gluster-vnzql Command: mkdir /var/lib/heketi/mounts/vg_548f/brick_8f08
Result:
[kubeexec] src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker7 Pod: gluster-vnzql Command: chown :2001 /var/lib/heketi/mounts/vg_548f/brick_8f08
[kubeexec] src/github.com/heketi/heketi/executors/kubeexec/kubexec.go:244: Host: worker7 Pod: gluster-vnzql Command: chmod 2775 /var/lib/heketi/mounts/vg_548f/brick_8f08
```

And I would expect to see these steps next:
```
[cmdexe] INFO Creating volume vol_<id> replica 3
[kubeexec] Debug <statement with the stuff that does the gluster command> to "gluster --mode=script volume create vol_<id> replica 3 <hosts of bricks it is mounting> 
Result: volume create: vol_<id> success: please start volume to access the data"
[kubeexec] <debug statement to the host that has the started volume on it> gluster --mode=script volume start vol_<id> 
Result: volume start: vol_<id>: success
[asynchttp] INFO <stuff here> Completed job <job_id> in ~7 seconds
```

But they are never happening and heketi just stops doing stuff, no error messages to be seen, no retries but it continues to heartbeat the volumes it has already created.

**Comment by phlogistonjohn** [comment]
As of version 6 Heketi generally operates in the following manner:
- The request is processed, the needed elements (devices, etc) are determined, and that is all logged in the db
- Heketi creates (or deletes, etc) the items from the gluster nodes.
- If no errors occurred the db entries are marked as no longer pending, otherwise attempts are made to undo the operation

Because of this there are a few common ways it can get out sync with the underlying systems.
1. Heketi process is terminated while it is performing an operation. This can be triggered by a node restart, pod move, etc.
2. Heketi is executing a command on a node and the communication channel fails in a way that the operation is unable to fail (see below). Typically this would eventually lead into the first item because the server with a blocked operation is eventually going to get restarted.
3. The gluster nodes that Heketi is working with are being restarted in such a way that the rollback can not complete. An example is a gluster node being restarted between creating a volume and starting the volume. This will fail the create operation but if there are too few nodes to allow deleting the incomplete volume heketi can not do anything more about the situation.

The 2nd point has been discussed recently among the Heketi developers. There is some concern that the kubernetes executor is not as stable as we need and that communicating with the gluster pods through that channel could lead to server hangs.

In cases 1 & 2 the things that heketi is going to process is logged in the db. These items can be viewed from a db dump (`heketi-cli db dump`) and can be manipulated via the offline command `heketi db delete-pending-entries` (please review the help prior to use).
You can also edit and reload the db dump manually but beware that this can lead to internal inconsistencies between the device storage counts if you add/remove bricks without making the correct changes to the device storage counters.

Case #3 was discovered after the version 6 release, AFAIR. There should be a fix in master for that condition. It will make #3 behave like 1 & 2 in that if the volume can not be cleaned up it will remain in the db for manual cleanup later.

If your shell sessions into the heketi container are being interrupted please verify that the pod/containter is being restarted. It may be failing liveness checks (due the root issue) and then being restarted by kubernetes. I'm not entirely sure about that part myself.

To wrap up, I think you are running into the following situation: something is preventing heketi from running operations to successful completion. This then may be causing heketi to be restarted. Restarts lead to stale operations and apparent de-syncs between heketi and the underlying system. While I can't guarantee this is the case it seems like a good place to start.

I would:
- Keep track of when / how often the heketi container get restarted
- Keep track of when / how often the gluster containers get restarted
- Monitor how busy the master node(s) are, I've been told that the kubernetes executor has to go through the master
- Use the `heketi-cli db dump` and `heketi db export` commands to check for pending operations in the db (the latter is better but must be done when heketi is stopped)
- Read up on the `heketi db delete-pending-entries` command to see if will help you resolve sync issues
- Attempt to root-cause the freezes you seem to experiencing. This may require you to disable liveness checks on the pod.

I know this is a lot to digest but without specifics all I can do it outline the general areas I would look into if I were debugging this issue.

**Comment by mjschmidt** [comment]
I have seen gluster node restarts that think may have coincided with heketi failure. I am not sure if this also causes heketi restarts or not. I will look into this more.

So I tore it down again for more testing. But I will try to figure out if it is 1 2 or 3 and get back to you all at some point. I appreciate the help.

**Comment by mjschmidt** [comment]
So my current way forward is to deploy gluster underneath kubernetes rather than in Kubernetes, and only run the heketi container inside of kubernetes since it seems like gluster container restarts are the issue I am facing. I want to see if deploying gluster not in containers solves my problems or not and debug from there, gluster is already distributed so I don't really need it running in a container.

@jarrpa if I am only running the heketi in the container for the deployment. What steps with gluster do I need to take ahead of time. Right now I have gluster installed (4.0 for now) on centos7 machines
yum list installed | grep gluster gives me
centos-release-gluuster40..x86_64
glusterfs.x86_64
glusterfs-api.x86_64
glusterfs-cli.x86_64
glusterfs-client-xlators.x86_64
glusterfs-fuse.x86_64
glusterfs-libs.x86_64
glusterfs-server.x86_64
userspace-rcu.x86_64

The gluster nodes are not connected together yet with peer probe yet. And they have no volumes attached to them.

**Comment by mjschmidt** [comment]
Also I spread the cluster out (less space on each node, but more nodes for same total amount of space)  to try to eliminate issue #3 that @phlogistonjohn pointed out.

**Comment by jarrpa** [comment]
You do not want to do a peer probe. You only want the glusterd server running, no configuration done. Also make sure you have the requisite kernel modules loaded, the needed firewall ports opened, and the target devices are all bare (`wipefs -a <device>` should do it).

**Comment by mjschmidt** [comment]
Okay that is easy (: I can uninstall and reinstall just to be safe on that front

**Comment by mjschmidt** [comment]
Ugh still working this. Had fips mode on... I don't think this is fully fixed in gluster 4

**Comment by jarrpa** [comment]
Agreed, that is a very sore point in the user experience. :(

What's the status at the moment? Have we really not gotten past initial deployment?

**Comment by mjschmidt** [comment]
So I have gotten past initial deployment both in containers and outside of containers, the issue is that I haven't been able to get any semblance of stability from Heketi and Gluster thus far.

Here my user evaluation thus far. If this is due to gaps in my understanding I welcome additional feedback and knowledge that can help me achieve what I am going for, which ultimately would be "cloud native storage for my k8s applications", though I would very much settle for performant/stable storage for my cloud native applications. This is why I wanted to go the Heketi/Gluster route in the first place.

I have been trying to test further with hard installed Heketi and Gluster providing storage up to Kubernetes on bare metal  servers due to @phlogistonjohn concerns with the kube-system pods not being stable enough for Heketi's purposes running in a container on Kubernetes. I was hoping the more external pieces I cut out the closer I would get to the root of the problems I am having and ultimately a final solution. So with hard install setup, I am seeing this crazy thing happen with Gluster where it crashes the machines and containers because for some unknown reason it is spinning up hundreds or thousands of Gluster deamons eventually crashing the machine. I can only assume that this same thing is happening in containerized gluster, though I have no hard evidence of this. This is when the real problems start as it causes to get Heketi out of sync with the cluster, its extremely difficult to clean up Gluster volumes, I also can't remove and re-provision problem nodes and get that back in sync with Heketi due to stuff discussed above, I can't force Heketi to remove volumes or nodes if dead nodes aren't working, and even if I could, I don't think I have the ability to say "okay as a system since I know that force removed node a had volumes 1 and 2 on it, I now need to redistribute that workload to other parts of the cluster". This functionality works, but only if the cluster is working perfectly and healthy, which I would argue that this functionality is desired mostly when the cluster is unhealthy and unhappy and a sys admin is attempting to get the cluster back to operating status.

I understand that these are all extremely difficult things to implement and there is a large amount of complexity with this sort of setup. You guys have been extremely helpful in getting me this far, but since we were talking status and where we are at. I definitely am not meaning to bash you guys in any way shape or form.

**Comment by jarrpa** [comment]
Thank you for the thorough feedback! Indeed, the failure scenario user experience leaves much to be desired. Howeverm I'll mention that the latest versions of heketi have the ability to directly edit their DB. While I don't know much about it personally, hopefully @phlogistonjohn can chime in with more.

**Comment by mjschmidt** [comment]
I mean that would be great because then I could go back and at least edit the db in heketi and have it working.

But that would basically make the memory leak issues that heketi did so much to fix gluster's problem because if I can't go clean up the gluster volumes (since a node is down) then still have a hard time figuring out how to use this in a prod setting? Suggestions on that front?

**Comment by mjschmidt** [comment]
So you guys over there really don't see any of these issues running gluster and heketi in contianers?

What OS are you guys running? What does your k8s cluster look like? I think you all mentioned you were bare metal? What is doing the install of Kubernetes? What version of K8s? So many questions.

**Comment by jarrpa** [comment]
Sorry for the delay, I'm been traveling and later swamped because of it.

Unfortunately, yes, we have really not seen issue like the ones you're running into. This is across the Red Hat family of distributions (Fedora, RHEL, etc.) in multiple environments including baremetal, VMs, and cloud. I myself have been running this since Kubernetes 1.7 on through 1.10, installed via kubeadm, on CentOS VMs.

**Comment by mjschmidt** [comment]
I think at this point we can close this issue. We have decided where we are to go the local persistent volume route and deal with managing the applications that need volumes manually until a better solution for local pvcs come out. A lot of applications in our use case are already distributed so we don't really use the gluster replication. I really appreciate the help and I learned a ton about volume management, gluster, heketi, containers and kubernetes.

**Comment by jarrpa** [comment]
No problem! It's been a wild ride, sorry that we couldn't make it work out for you, but I hope the local storage solution proves sufficient!
